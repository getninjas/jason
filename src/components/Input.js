import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import IMask from 'imask';
import { maxLengthTrim, getInputType } from '../helpers/input';

const propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onFieldBlur: PropTypes.func.isRequired,
  title: PropTypes.string,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  initialValue: PropTypes.any,
  values: PropTypes.array,
  style: PropTypes.string,
  minLength: PropTypes.number,
  maxLength: PropTypes.number,
};

const defaultProps = {
  placeholder: '',
  required: false,
  title: '',
  type: 'text',
  initialValue: '',
  values: [],
  style: 'form__input',
  minLength: 3,
  maxLength: 255,
};

export default class Input extends Component {
  constructor(props) {
    super(props);

    this.state = {
      values: this.props.values,
      value: this.props.initialValue ? this.props.initialValue : '',
    };

    this.ref = createRef();
    this.onChange = this.onChange.bind(this);
    this.onChangeCheck = this.onChangeCheck.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.mask = null;
  }

  onChange(evt) {
    const inputValue = maxLengthTrim(evt.target.value, this.props.maxLength);

    this.props.onFieldChange({
      value: inputValue,
      id: this.props.id,
    });

    this.setState({ value: inputValue });
  }

  onChangeCheck(evt) {
    console.log('evt', evt);
    let idList = [];
    if (!this.state.value.length) {
      idList.push(parseInt(evt.target.id, 10));
    } else {
      idList = this.state.value.filter(valor => valor !== parseInt(evt.target.id, 10));

      if (idList.length === this.state.value.length) {
        idList.push(parseInt(evt.target.id, 10));
      }
    }

    this.props.onFieldChange({
      id: this.props.id,
      value: idList,
    });

    this.setState({ value: idList });
  }

  onBlur(evt) {
    this.props.onFieldBlur({
      value: evt.target.value,
      id: this.props.id,
      required: this.props.required,
      type: this.props.type,
      minLength: this.props.minLength,
    });
  }

  componentDidMount() {
    const { type } = this.props;

    if (type === 'phone') {
      this.mask = new IMask(this.ref.current, { mask: '(00) 00000-0000' });
      this.mask.on('complete', () => {
        this.props.onFieldChange({
          value: this.mask.value,
          id: this.props.id,
        });

        this.setState({ value: this.mask.value });
      });
    }
  }

  render() {
    const {
      type,
      id,
      name,
      title,
      style,
      placeholder,
      required,
      minLength,
      maxLength,
    } = this.props;

    const prepareComponent = () => {
      if (type === 'checkbox') {
        return (
          <div required={required ? 'true' : 'false'}>
            {this.state.values.map((elem, idx) => (
                <label key={`${elem.databaseId}-${idx}`} htmlFor={elem.databaseId}>
                  <input
                    type={getInputType(type)}
                    id={elem.databaseId}
                    name={name}
                    title={title}
                    className={style}
                    value={elem.value}
                    onChange={this.onChangeCheck}
                    ref={this.ref} />
                    {elem.value === 'OTHER' ? (
                      <input type="text" />
                    ) : elem.value}
                </label>
              ))}
          </div>
        );
      }

      return (
        <input
          type={getInputType(type)}
          id={id}
          name={name}
          title={title}
          className={style}
          placeholder={placeholder}
          required={required ? 'true' : 'false'}
          value={this.state.value}
          onChange={this.onChange}
          onBlur={this.onBlur}
          minLength={minLength}
          maxLength={maxLength}
          ref={this.ref} />
      );
    };

    return prepareComponent();
  }
}

Input.propTypes = propTypes;
Input.defaultProps = defaultProps;
