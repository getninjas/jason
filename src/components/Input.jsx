import React from 'react';
import PropTypes from 'prop-types';

export default class Input extends React.Component {
  constructor() {
    super();

    this.state = {
      value: '',
    }

    this.onChange = this.onChange.bind(this);
  }

  onChange(evt) {
    this.setState({ value: evt.target.value });
  }

  componentWillMount() {
    this.setState({ value: this.props.value });
  }

  render() {
    return (
      <input
        type={this.props.type==='phone'? 'tel': this.props.type}
        id={this.props.id}
        name={this.props.name}
        title={this.props.title}
        className="form__input"
        placeholder={this.props.placeholder}
        required={this.props.required ? 'true' : 'false'}
        value={this.state.value}
        onChange={this.onChange} />
    );
  }
}

Input.defaultProps = {
  id: '',
  placeholder: '',
  required: false,
  name: '',
  type: 'text',
  title: '',
  value: '',
}

Input.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  value: PropTypes.any,
}
