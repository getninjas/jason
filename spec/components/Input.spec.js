import React from 'react';
import renderer from 'react-test-renderer';
import Input from '../../src/components/Input';
import { enzymeConfig, shallow, mount } from '../enzymeConfig';

enzymeConfig();

function createNodeMock(element) {
  if (element.type === 'input') {
    return {
      addEventListener() {},
      value: '',
    };
  }

  return null;
}

describe('Input', () => {
  const commonProps = {
    id: 'idTest',
    name: 'nameTest',
    onFieldChange: () => {},
    onFieldBlur: () => {},
    placeholder: 'placeholderTest',
    required: false,
  };

  it('renders defaultProps', () => {
    const options = { createNodeMock };
    const component = renderer.create(
      <Input
        {...commonProps}
        value={'ola test value'}
      />, options,
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('triggers onFieldChange', () => {
    const onFieldChange = jest.fn();

    const component = shallow(
      <Input
        id={'idTest'}
        name={'nameTest'}
        onFieldBlur={() => {}}
        onFieldChange={onFieldChange}
        placeholder={'placeholderTest'}
        required={false}
        value={'ola test value'}
      />,
    );

    component.simulate('change', { target: { value: 'Bora pra action' } });

    expect(component.instance().props.onFieldChange).toBeCalledWith({
      value: 'Bora pra action',
      id: 'idTest',
    });
  });

  it('trims text to maxLength', () => {
    const onFieldChange = jest.fn();

    const component = mount(
      <Input
        {...commonProps}
        id={'idTest'}
        onFieldChange={onFieldChange}
        value={''}
        maxLength={5}
      />,
    );

    component.simulate('change', { target: { value: 'Bora pra action' } });

    expect(component.instance().props.onFieldChange).toBeCalledWith({
      value: 'Bora ',
      id: 'idTest',
    });
  });

  describe('with type', () => {
    it('renders type text', () => {
      const component = shallow(<Input id={'id_input'} name={'input_name'} onFieldChange={() => {}} onFieldBlur={() => {}} />);

      expect(component.prop('type')).toBe('text');
    });

    it('renders type email', () => {
      const component = shallow(<Input id={'id_input'} name={'input_name'} onFieldChange={() => {}} onFieldBlur={() => {}} type={'email'}/>);

      expect(component.prop('type')).toBe('email');
    });

    it('renders type tel when type equal phone', () => {
      const component = mount(<Input id={'id_input'} name={'input_name'} onFieldChange={() => {}} onFieldBlur={() => {}} type={'phone'}/>);

      const inputType = component.getDOMNode().attributes.type.value;

      expect(inputType).toBe('tel');
    });
  });

  describe('.onBlur', () => {
    it('calls onFieldBlur', () => {
      const onFieldBlur = jest.fn();

      const component = shallow(
        <Input
          id={'idTest'}
          name={'nameTest'}
          onFieldBlur={onFieldBlur}
          onFieldChange={() => {}}
          placeholder={'placeholderTest'}
          required={false}
          value={''}
          maxLength={5}
        />,
      );

      component.simulate('blur', { target: { value: '(11) 99999-8888' } });

      expect(component.instance().props.onFieldBlur).toBeCalled();
    });

    describe('when input is type phone', () => {
      it('triggers onFieldBlur with target values', () => {
        const onFieldBlur = jest.fn();

        const component = mount(
          <Input
            id='phone'
            name='phone'
            type='phone'
            onFieldChange={() => {}}
            onFieldBlur={onFieldBlur}
            placeholder='(__) _____-____'
            required={false}
            value=''
          />,
        );

        component.simulate('change', { target: { value: '(11) 99999-8888' } });
        component.simulate('blur', { target: { value: '(11) 97878-1212' } });

        expect(component.instance().props.onFieldBlur).toBeCalledWith({
          id: 'phone',
          minLength: 1,
          required: false,
          type: 'phone',
          value: '(11) 97878-1212',
        });
      });
    });
  });

  describe('.onChange', () => {
    it('applies mask input', () => {
      const onFieldChange = jest.fn();

      const component = mount(
        <Input
          id='phone'
          name='phone'
          type='phone'
          onFieldChange={onFieldChange}
          onFieldBlur={() => {}}
          placeholder='(__) _____-____'
          required={false}
          value=''
        />,
      );

      const instance = component.instance();

      component.simulate('change', { target: { value: '11999998888' } });

      expect(instance.props.onFieldChange).toBeCalledWith({
        id: 'phone',
        value: '(11) 99999-8888',
      });
    });

    it('calls applyMask onChange', () => {
      const component = mount(
        <Input
          id='phone'
          name='phone'
          type='phone'
          onFieldChange={() => {}}
          onFieldBlur={() => {}}
          placeholder='(__) _____-____'
          required={false}
          value=''
        />,
      );

      const instance = component.instance();

      instance.applyMask = jest.fn();

      component.simulate('change', { target: { value: '11999998888' } });

      expect(instance.applyMask).toBeCalledWith('11999998888');
    });
  });

  describe('.applyMask', () => {
    describe('when input type phone', () => {
      it('returns masked value', () => {
        const component = mount(
          <Input
            id='phone'
            name='phone'
            type='phone'
            onFieldChange={() => {}}
            onFieldBlur={() => {}}
            placeholder='(__) _____-____'
            required={false}
            value=''
          />,
        );

        expect(component.instance().applyMask('11999998888')).toBe('(11) 99999-8888');
      });
    });

    describe('when input not of type phone', () => {
      it('does not apply mask', () => {
        const component = mount(
          <Input
            id='idTest'
            name='idTest'
            type='text'
            onFieldChange={() => {}}
            onFieldBlur={() => {}}
            placeholder='(__) _____-____'
            required={false}
            value=''
          />,
        );

        expect(component.instance().applyMask('John Doe')).toBe('John Doe');
      });
    });
  });
});
