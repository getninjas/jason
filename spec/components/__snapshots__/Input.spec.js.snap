// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders defaultProps 1`] = `
<input
  className="form__input"
  id="idTest"
  maxLength={255}
  minLength={1}
  name="nameTest"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="placeholderTest"
  required="false"
  title=""
  type="text"
  value=""
/>
`;
